File: Febrero2019/DataStructures/Graph/DictionaryWeightedGraph.hs
----------------------------------------------
-- Estructuras de Datos.  2018/19
-- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
-- Escuela Técnica Superior de Ingeniería en Informática. UMA
--
-- Examen 4 de febrero de 2019
--
-- ALUMNO/NAME:
-- GRADO/STUDIES:
-- NÚM. MÁQUINA/MACHINE NUMBER:
--
-- Weighted Graph implemented by using a dictionary from
-- sources to another dictionary from destinations to weights
----------------------------------------------

module DataStructures.Graph.DictionaryWeightedGraph
  ( WeightedGraph
  , WeightedEdge(WE)
  , empty
  , isEmpty
  , mkWeightedGraphEdges
  , addVertex
  , addEdge
  , vertices
  , numVertices
  , edges
  , numEdges
  , successors
  ) where

import Data.List(nub, intercalate)

import Data.Maybe

import qualified DataStructures.Dictionary.AVLDictionary as D

data WeightedEdge a w  = WE a w a deriving Show

instance (Eq a, Eq w) => Eq (WeightedEdge a w) where
  WE u w v == WE u' w' v' = (u==u' && v==v' || u==v' && v==u')
                              && w == w'

instance (Eq a, Ord w) => Ord (WeightedEdge a w) where
  compare (WE _ w _) (WE _ w' _) = compare w w'

data WeightedGraph a w  = WG (D.Dictionary a (D.Dictionary a w))

empty :: WeightedGraph a w
empty = WG D.empty

addVertex :: (Ord a) => WeightedGraph a w -> a -> WeightedGraph a w
addVertex (WG dict) e = WG (D.insert e (D.empty) dict)

addEdge :: (Ord a, Show a) => WeightedGraph a w -> a -> a -> w -> WeightedGraph a w
addEdge (WG dict) s d w 
  | (D.isDefinedAt s dict) && (D.isDefinedAt d dict) = WG (D.insert s (D.insert d w (fromJust(D.valueOf s dict))) dict)
  | otherwise =  error "vertices no definidos"
  

edges :: (Eq a, Eq w) => WeightedGraph a w -> [WeightedEdge a w]
edges  (WG dict) = aux (D.keysValues dict) [] 
  where
    aux [] lista = lista
    aux ((s,dw):xs) lista = aux xs ([WE s w d | (d,w) <- D.keysValues dw] ++ lista)

successors :: (Ord a, Show a) => WeightedGraph a w -> a -> [(a,w)]
successors (WG dict) v
  | (D.isDefinedAt v dict) = aux [] (D.keysValues (fromJust(D.valueOf v dict))) v
  | otherwise = error "vertices no definidos"
    where
      aux lista [] v = lista
      aux lista ((dst,w):xs) v = aux ([(dst,w)] ++ lista) xs v 






-- NO EDITAR A PARTIR DE AQUÍ    
-- DON'T EDIT ANYTHING BELOW THIS COMMENT

vertices :: WeightedGraph a w -> [a]
vertices (WG d) = D.keys d

isEmpty :: WeightedGraph a w -> Bool
isEmpty (WG d) = D.isEmpty d

mkWeightedGraphEdges :: (Ord a, Show a) => [a] -> [WeightedEdge a w] -> WeightedGraph a w
mkWeightedGraphEdges vs es = wg'
  where
    wg = foldl addVertex empty vs
    wg' = foldr (\(WE u w v) wg -> addEdge wg u v w) wg es

numVertices :: WeightedGraph a w -> Int
numVertices = length . vertices

numEdges :: (Eq a, Eq w) => WeightedGraph a w -> Int
numEdges = length . edges

instance (Eq a, Show a, Eq w, Show w) => Show (WeightedGraph a w) where
  show wg  = "DictionaryWeightedGraph("++vs++", "++as++")"
   where
    vs  = "("++ intercalate ", " (map show (vertices wg)) ++")"
    as  = "(" ++ intercalate ", " (map showEdge (edges wg)) ++ ")"
    showEdge (WE x w y)  = intercalate "-" [ show x, show w, show y ]

File: Febrero2019/feb2019java/src/dataStructures/graph/DictionaryWeightedGraph.java
/**----------------------------------------------
 * -- Estructuras de Datos.  2018/19
 * -- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
 * -- Escuela Técnica Superior de Ingeniería en Informática. UMA
 * --
 * -- Examen 4 de febrero de 2019
 * --
 * -- ALUMNO/NAME:
 * -- GRADO/STUDIES:
 * -- NÚM. MÁQUINA/MACHINE NUMBER:
 * --
 * ----------------------------------------------
 */

package dataStructures.graph;

import java.util.Iterator;

import dataStructures.dictionary.Dictionary;
import dataStructures.dictionary.HashDictionary;

import dataStructures.set.Set;
import dataStructures.set.HashSet;
import dataStructures.tuple.Tuple2;

import javax.xml.transform.SourceLocator;

public class DictionaryWeightedGraph<V, W extends Comparable<? super W>> implements WeightedGraph<V, W> {

    static class WE<V1, W1 extends Comparable<? super W1>> implements WeightedEdge<V1, W1> {

		V1 src, dst;
        W1 wght;

        WE(V1 s, V1 d, W1 w) {
            src = s;
            dst = d;
            wght = w;
        }

        public V1 source() {
            return src;
        }

        public V1 destination() {
            return dst;
        }

        public W1 weight() {
            return wght;
        }

        public String toString() {
            return "WE(" + src + "," + dst + "," + wght + ")";
        }

		public int hashCode() {
			
			return src.hashCode() + dst.hashCode() + wght.hashCode();
		}

		public boolean equals(Object obj) {
			boolean ok = false;
            if(obj instanceof WE){
                WE other = (WE) obj;
                ok = this.src.equals(other.src) && this.dst.equals(other.dst) && this.wght.equals(other.wght);
            }
            return ok;
		}
		
		public int compareTo(WeightedEdge<V1, W1> o) {
			return wght.compareTo(o.weight());
		}
    }

    /**
     * Each vertex is associated to a dictionary containing associations
     * from each successor to its weight
     */
    protected Dictionary<V, Dictionary<V, W>> graph;

    public DictionaryWeightedGraph() {
        graph = new HashDictionary<>();
    }


    public void addVertex(V v) {
    	graph.insert(v,new HashDictionary<>());
    }

    public void addEdge(V src, V dst, W w) {
        if(graph.isDefinedAt(src) && graph.isDefinedAt(dst)){
            graph.valueOf(src).insert(dst,w);
        }else {
            throw new GraphException("Vertices no definidos");
        }
    }

    public Set<Tuple2<V, W>> successors(V v) {
        if(graph.isDefinedAt(v)){
            Set<Tuple2<V, W>> sol = new HashSet<>();
            for(Tuple2<V,W> t:graph.valueOf(v).keysValues()){
                sol.insert(t);
            }
            return sol;
        }else{
            throw new GraphException("Vertice no definidos");
        }
    }


    public Set<WeightedEdge<V, W>> edges() {
        Set<WeightedEdge<V, W>> sol = new HashSet<>();
        for(V v : graph.keys()){
            for(Tuple2<V,W> t:successors(v)){
                sol.insert(new WE<>(v,t._1(),t._2()));
            }
        }

        return sol;
    }


    /** DON'T EDIT ANYTHING BELOW THIS COMMENT **/


    public Set<V> vertices() {
        Set<V> vs = new HashSet<>();
        for (V v : graph.keys())
            vs.insert(v);
        return vs;
    }


    public boolean isEmpty() {
        return graph.isEmpty();
    }

    public int numVertices() {
        return graph.size();
    }


    public int numEdges() {
        int num = 0;
        for (Dictionary<V, W> d : graph.values())
            num += d.size();
        return num;
    }


    public String toString() {
        String className = getClass().getSimpleName();
        String s = className + "(vertices=(";

        Iterator<V> it1 = vertices().iterator();
        while (it1.hasNext())
            s += it1.next() + (it1.hasNext() ? ", " : "");
        s += ")";

        s += ", edges=(";
        Iterator<WeightedEdge<V, W>> it2 = edges().iterator();
        while (it2.hasNext())
            s += it2.next() + (it2.hasNext() ? ", " : "");
        s += "))";

        return s;
    }
}
